modules/reactor/src/meshgenerators/CoreMeshGenerator.C

```diff
@@ -0,0 +1,402 @@
+
+  params.addRequiredParam<MeshGeneratorName>(
+      "dummy_assembly_name",
+      "The place holder name in \"inputs\" that indicates an empty position.");
+
+  params.addRequiredParam<std::vector<std::vector<unsigned int>>>(
+      "pattern", "A double-indexed array starting with the upper-left corner");
+
+  params.addParam<bool>("extrude",
+                        false,
+                        "Determines if this is the final step in the geometry construction"
+                        " and extrudes the 2D geometry to 3D. If this is true then this mesh "
+                        "cannot be used in further mesh building");
+
+  params.addClassDescription("This CoreMeshGenerator object is designed to generate a core-like "
+                             "structure, with IDs, from a reactor geometry. "
+                             "The core-like structure consists of a pattern of assembly-like "
+                             "structures generated with AssemblyMeshGenerator "
+                             "and is permitted to have \"empty\" locations. The size and spacing "
+                             "of the assembly-like structures is defined, and "
+                             "enforced by declaration in the ReactorMeshParams.");
+
+  return params;
+}
+
+CoreMeshGenerator::CoreMeshGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _inputs(getParam<std::vector<MeshGeneratorName>>("inputs")),
+    _empty_key(getParam<MeshGeneratorName>("dummy_assembly_name")),
+    _pattern(getParam<std::vector<std::vector<unsigned int>>>("pattern")),
+    _extrude(getParam<bool>("extrude"))
+{
+  /* Below, we get the ReactorMeshParams info from one of the pin inputs.
+     The reactor params gives the axial info if we are to extrude as well as the assembly pitch
+     that the pin pattern is to be checked against.
+     We then create the PatternedMesh subgenerators and pass the parameter info needed for the
+     inputs. Finally, if the the mesh is to be extruded we create the FancyExtruderGenerator
+     subgenerator and the PlaneIDMeshGenerator subgenerator to handle this.
+  */
+  MeshGeneratorName _reactor_params =
+      MeshGeneratorName(getMeshProperty<std::string>("reactor_params_name", _inputs[0]));
+
+  // Ensure that the user has supplied the correct info for conformal mesh generation
+  if (!hasMeshProperty("mesh_dimensions", _reactor_params) ||
+      !hasMeshProperty("mesh_geometry", _reactor_params))
+    mooseError("The reactor_params input must be a ReactorMeshParams type MeshGenerator\n");
+  else
+  {
+    _geom_type = getMeshProperty<std::string>("mesh_geometry", _reactor_params);
+    _mesh_dimensions = getMeshProperty<int>("mesh_dimensions", _reactor_params);
+  }
+  if (_extrude && getMeshProperty<unsigned int>("mesh_dimensions", _reactor_params) != 3)
+    mooseError("This is a 2 dimensional mesh, you cannot extrude it. Check your ReactorMeshParams "
```

Comment:
see above. I wonder if there might be enough shared code to warrant a common base class.

Reply:
After looking this over, it doesn't seem like there is a great deal of overlap between PinMeshGenerator, AssemblyMeshGenerator, and CoreMeshGenerator besides the parameter `_extrude` and the check above to see if the option to extrude the mesh is valid. Not sure if this is enough to warrant a common base class